### PR TITLE
Add Space chat/cc crawler

### PR DESCRIPTION
**Pull request by HitomaruKonpaku**

### CODE REVIEW

File: src/SpaceChat.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  private ws: WebSocket
+  private accessChatData: Record<string, any>
+
+  private tmpChatFile: string
+  private outChatFile: string
+
+  constructor(
+    public spaceId: string,
+    private liveStreamStatus: Record<string, any>,
+    private options: Record<string, any>,
```
Comment (IAmVisco) [inline src/SpaceChat.ts:25]: What are options? Might wanna type them.

File: src/SpaceChat.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  ) {
+    this.logger = baseLogger.child({ label: `[SpaceChat@${spaceId}]` })
+    this.spaceId = spaceId
+    this.liveStreamStatus = liveStreamStatus
+    this.options = options
```
Comment (IAmVisco) [inline src/SpaceChat.ts:30]: Setting `private` in the constructor already makes them fields. No need to do this again.

File: src/SpaceChat.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    this.tmpChatFile = path.join(Downloader.getMediaDir(this.username), `${this.filename} Chat.jsonl`)
+    this.outChatFile = path.join(Downloader.getMediaDir(this.username), `${this.filename} Chat.txt`)
+
+    Downloader.createMediaDir(this.username)
```
Comment (IAmVisco) [inline src/SpaceChat.ts:35]: This sounds like an Util method.

File: src/SpaceChat.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    this.logger.info(`Tmp chat file ${this.tmpChatFile}`)
```
Comment (IAmVisco) [inline src/SpaceChat.ts:36]: Assignment returns value, you can log it above (but it's rarely used).

File: src/SpaceChat.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    this.logger.info(`Out chat file ${this.outChatFile}`)
+  }
+
+  private get chatToken(): string {
+    return this.liveStreamStatus.chatToken
+  }
+
+  private get username(): string {
+    return this.options?.username || ''
```
Comment (IAmVisco) [inline src/SpaceChat.ts:45]: Types don't indicate that `options` are optional.

File: src/SpaceChat.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    this.logger.debug('accessChat data', data)
+    this.accessChatData = data
+  }
+
+  private initWebSocket() {
+    this.ws = new WebSocket(this.WS_URL, {})
+    this.attachWsEvents()
+  }
+
+  private attachWsEvents() {
```
Comment (IAmVisco) [inline src/SpaceChat.ts:82]: More of `setupWsEventHandlers` or along the lines.

File: src/SpaceChat.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    const { ws } = this
+    ws.on('error', (error) => {
+      this.logger.error(`[WS] Error: ${error.message}`)
```
Comment (IAmVisco) [inline src/SpaceChat.ts:85]: Can make another child logger in this scope I guess.

File: src/SpaceChat.ts [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    }
+  }
+
+  private async processTmpChatFile() {
+    this.logger.silly('processTmpChatFile')
+    const fileStream = fs.createReadStream(this.tmpChatFile)
+    const rl = readline.createInterface({ input: fileStream })
+    let lineCount = 0
+    // eslint-disable-next-line no-restricted-syntax
+    for await (const line of rl) {
```
Comment (IAmVisco) [inline src/SpaceChat.ts:136]: `for await` is used for **generators**. You don't need it here.

File: src/SpaceChat.ts [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+      lineCount += 1
```
Comment (IAmVisco) [inline src/SpaceChat.ts:137]: You can just use `Array.forEach()` here tbh.

File: src/SpaceChat.ts [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  private processChatBody(payload: string) {
+    const obj = JSON.parse(payload)
+    if (!obj.final || !obj.body) {
+      return
+    }
+    const msg = `${obj.username}: ${obj.body}\n`
+    fs.appendFileSync(this.outChatFile, msg)
+  }
```
Comment (IAmVisco) [inline src/SpaceChat.ts:170]: Not sure if you need to split it that much but your call.

File: src/SpaceWatcher.ts [hunk 11]
```diff
@@ -132,6 +153,21 @@ export class SpaceWatcher extends EventEmitter {
     setTimeout(() => this.downloadMedia(), timeoutMs)
   }
 
+  private watchSpaceChat() {
```
Comment (IAmVisco) [inline src/SpaceWatcher.ts:156]: Also it's not `chat`. It's transcription or closed captions.

File: src/enums/SpaceChatKind.enum.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+/* eslint-disable no-unused-vars */
+export enum SpaceChatKind {
```
Comment (IAmVisco) [inline src/enums/SpaceChatKind.enum.ts:2]: I think enums are styled in SCREAMING_SNAKE_CASE in singular.

**Review by IAmVisco — CHANGES_REQUESTED** [review]

File: src/SpaceChat.ts (hunk 7 above)
Comment (HitomaruKonpaku) [reply src/SpaceChat.ts]: For now nothing much here so prob dont need 🤔

File: src/SpaceWatcher.ts [hunk 13]
```diff
@@ -10,18 +10,19 @@ import { APP_PLAYLIST_REFRESH_INTERVAL } from './constants/app.constant'
...
-import { SpaceChat } from './SpaceChat'
+import { Twitter } from './namespaces/Twitter'
+import { SpaceCaptions } from './SpaceCaptions'
 import { Util } from './Util'
 
 export class SpaceWatcher extends EventEmitter {
   private logger: winston.Logger
-  private metadata: Record<string, any>
-  private liveStreamStatus: Record<string, any>
+  private metadata: Twitter.AudioSpaceMetadata
```
Comment (IAmVisco) [inline src/SpaceWatcher.ts:19]: If you are exporting it from a file named `Twitter` you should not have extra `Twitter` object here.

File: src/SpaceWatcher.ts [hunk 14]
```diff
@@ -10,18 +10,19 @@ import { APP_PLAYLIST_REFRESH_INTERVAL } from './constants/app.constant'
 import { TWITTER_AUTHORIZATION } from './constants/twitter.constant'
 import { Downloader } from './Downloader'
 import { logger as baseLogger } from './logger'
-import { SpaceChat } from './SpaceChat'
+import { Twitter } from './namespaces/Twitter'
+import { SpaceCaptions } from './SpaceCaptions'
```
Comment (IAmVisco) [inline src/SpaceWatcher.ts:14]: Shouldn't this be in some `interfaces` folder or something

File: src/index.ts [hunk 15]
```diff
@@ -22,6 +23,15 @@ class Main {
       this.logger.info('Starting...')
       this.logger.info('Args', args)
 
+      if (typeof args['extract-cc'] === 'string') {
```
Comment (IAmVisco) [inline src/index.ts:26]: Is this how you parse args lol. Check `commander` package.

**Review by IAmVisco — CHANGES_REQUESTED** [review]

File: src/SpaceWatcher.ts (hunk 14 above)
Comment (HitomaruKonpaku) [reply src/SpaceWatcher.ts:14]: It includes interfaces/enums so better have separate folder 🤔

**Pull request merged**